### PR TITLE
Add support for unit-testing client callbacks on brokered services

### PR DESCRIPTION
**Pull request by murphymcquet-msft**
In the current `BrokeredServiceContractTestBase<TInterface, TServiceMock>` class, it's not possible to configure client callbacks on the `TServiceMock` that is provided. In an actual request for a brokered service, we may pass a `Type` to the `ServicerpcDescriptor.ClientInterface` which should provide a callback for the brokered service into the client, but as of today that functionality does not exist to unit test. This PR provides a mechanism to UT such scenarios where a client callback is required.

The main changes in this PR are:
1. Adjust the `InitializeAsync()` method of `BrokeredServiceContractTestBase<TInterface, TServiceMock>` so that it sets up RPC connections for the server and client, and then passes those connections to a `virtual` method which handles configuration of the service and client callback proxies. On the base class, set up the service proxy.
2. Create a new class `BrokeredServiceContractTestBase<TInterface, TServiceMock, TClientInterface, TClientMock>` that inherits from `BrokeredServiceContractTestBase<TInterface, TServiceMock>` . Override the method for configuring service and client callback proxies so that it will call the base implementation and then set up the client callback proxy.
3. Define a new interface `IMockServiceWithClientCallback` which defines a property `public object ClientInterface` with a get/set method. If a test class derives from `BrokeredServiceContractTestBase<TInterface, TServiceMock, TClientInterface, TClientMock>` then the type passed for `TServiceMock` must implement `IMockServiceWithClientCallback` and the type passed as the `ServiceRpcDescriptor.ClientInterface` must be an instance of `TClientMock`.

### CODE REVIEW

File: src/Microsoft.VisualStudio.Sdk.TestFramework.Xunit/BrokeredServiceContractTestBase`3.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+{
+    /// <summary>
+    /// A base class for testing Visual Studio brokered service contracts with client callbacks.
+    /// </summary>
+    /// <typeparam name="TInterface">The service interface.</typeparam>
+    /// <typeparam name="TServiceMock">The class that mocks the service.</typeparam>
+    /// <typeparam name="TClientInterfaceMock">The class that mockes the service's client callback.</typeparam>
+    public abstract class BrokeredServiceContractTestBase<TInterface, TServiceMock, TClientInterfaceMock> : BrokeredServiceContractTestBase<TInterface, TServiceMock>
+        where TInterface : class
+        where TServiceMock : TInterface, IMockServiceWithClientCallback, new()
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Sdk.TestFramework.Xunit/BrokeredServiceContractTestBase`3.cs:21]: 👏🏼 I like the additional type constraint you came up with here. That should help people realize how they need to work with this.

File: src/Microsoft.VisualStudio.Sdk.TestFramework.Xunit/BrokeredServiceContractTestBase`3.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            this.ClientInterface = new TClientInterfaceMock();
+        }
+
+        /// <summary>
+        /// Gets or sets the mock client callback instance.
+        /// </summary>
+        public TClientInterfaceMock ClientInterface { get; protected set; }
+
+        /// <inheritdoc/>
+        public override async Task InitializeAsync()
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Sdk.TestFramework.Xunit/BrokeredServiceContractTestBase`3.cs:47]: 👮🏼 This appears to set up an entirely distinct RPC stream for the callback interface. But that isn't how the product works. If we change this to better emulate the product, I think you'll find _much_ less code is required in this method. Or if you break up the base method as we discussed, you may not end up overridding this particular method at all.

File: src/Microsoft.VisualStudio.Sdk.TestFramework/IMockServiceWithClientCallback.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Sdk.TestFramework
+{
+    /// <summary>
+    /// Interface for service mocks that need to interact with a client callback proxy.
+    /// </summary>
+    public interface IMockServiceWithClientCallback
+    {
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Sdk.TestFramework/IMockServiceWithClientCallback.cs:16]: 💅🏼 Prefer namespace statements over namespace blocks for all new files.

**Review by AArnott — CHANGES_REQUESTED** [review]

File: src/Microsoft.VisualStudio.Sdk.TestFramework.Xunit/BrokeredServiceContractTestBase`3.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+{
+    /// <summary>
+    /// Initializes a new instance of the <see cref="BrokeredServiceContractTestBase{TInterface, TServiceMock, TClientInterfaceMock}"/> class.
+    /// </summary>
+    /// <param name="logger"><inheritdoc cref="LoggingTestBase(ITestOutputHelper)" path="/param[@name='logger']"/></param>
+    /// <param name="serviceRpcDescriptor">The descriptor that the product will use to request or proffer the brokered service.</param>
+    public BrokeredServiceContractTestBase(ITestOutputHelper logger, ServiceRpcDescriptor serviceRpcDescriptor)
+        : base(logger, serviceRpcDescriptor)
+    {
+        Assert.NotNull(serviceRpcDescriptor.ClientInterface);
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Sdk.TestFramework.Xunit/BrokeredServiceContractTestBase`3.cs]: This isn't the test itself. It's a test framework. Use `Requires` for argument validation rather than xunit's `Assert` class.

File: src/Microsoft.VisualStudio.Sdk.TestFramework.Xunit/BrokeredServiceContractTestBase`3.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        Assert.Equal(typeof(TClientInterfaceMock), serviceRpcDescriptor.ClientInterface.GetType());
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Sdk.TestFramework.Xunit/BrokeredServiceContractTestBase`3.cs]: I expect this to fail every time, because `serviceRpcDescriptor.ClientInterface.GetType()` will always be `typeof(Type)`.
The issue here is that `serviceRpcDescriptor.ClientInterface` itself is typed as `System.Type`, so you don't need to get its underlying Type or you'll miss the value you actually want to check.

**Review by AArnott — CHANGES_REQUESTED** [review]
This looks much better.
I pushed a commit with a few small improvements, but I left my messages active to call out the important changes.

The only thing lacking now are tests to verify the new base class.

File: src/Microsoft.VisualStudio.Sdk.TestFramework.Xunit/BrokeredServiceContractTestBase`3.cs (hunk 5 above)
Comment (murphymcquet-msft) [reply src/Microsoft.VisualStudio.Sdk.TestFramework.Xunit/BrokeredServiceContractTestBase`3.cs]: Yes, that was a miss. Thanks for making that change - I should have verified that the type was an implementation of the interface.

File: src/Microsoft.VisualStudio.Sdk.TestFramework.Xunit/BrokeredServiceContractTestBase`2.cs [hunk 6]
```diff
@@ -79,6 +78,11 @@ public BrokeredServiceContractTestBase(ITestOutputHelper logger, ServiceRpcDescr
     /// </summary>
     protected SourceLevels DescriptorLoggingVerbosity { get; set; } = SourceLevels.Verbose;
 
+    /// <summary>
+    /// Gets the test count.
+    /// </summary>
+    protected int TestCounter => testCounter;
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Sdk.TestFramework.Xunit/BrokeredServiceContractTestBase`2.cs:84]: Why was this necessary? It doesn't appear to be used anywhere.

File: test/Microsoft.VisualStudio.Sdk.TestFramework.Tests/Mocks/GreetMock.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+    public event EventHandler? OperationComplete;
+
+    public object ClientCallback { get; set; } = new object();
+
+    public async ValueTask<string> MakeGreeting(CancellationToken cancellationToken)
+    {
+#pragma warning disable CS8605
+        var name = await (ValueTask<string>)typeof(ISayName).GetMethod("SayName")?.Invoke(this.ClientCallback, null);
+#pragma warning restore CS8605
```
Comment (AArnott) [inline test/Microsoft.VisualStudio.Sdk.TestFramework.Tests/Mocks/GreetMock.cs:18]: Reflection isn't necessary here. Just a type cast.

```suggestion
        string name = await ((ISayName)this.ClientCallback).SayName();
```

File: test/Microsoft.VisualStudio.Sdk.TestFramework.Tests/Mocks/GreetMock.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+using static System.Reflection.BindingFlags;
+
+public class GreetMock : IGreet, IMockServiceWithClientCallback
+{
+    public event EventHandler<string>? GreetingSent;
+
+    public event EventHandler? OperationComplete;
+
+    public object ClientCallback { get; set; } = new object();
```
Comment (AArnott) [inline test/Microsoft.VisualStudio.Sdk.TestFramework.Tests/Mocks/GreetMock.cs:12]: nit: I don't think we should initialize this with an object that doesn't implement the expected interface.
But I suppose you're trying to avoid declaring the property as nullable, which makes sense since we guarantee that the base class will initialize it. So you could do this instead:

```suggestion
    public object ClientCallback { get; set; } = null!;
```

**Review by AArnott — APPROVED** [review]
Look great. Just a couple nits to solve.

File: src/Microsoft.VisualStudio.Sdk.TestFramework.Xunit/BrokeredServiceContractTestBase`2.cs (hunk 6 above)
Comment (murphymcquet-msft) [reply src/Microsoft.VisualStudio.Sdk.TestFramework.Xunit/BrokeredServiceContractTestBase`2.cs:84]: Artifact of a previous iteration, thanks for catching that. I'll remove it.

File: test/Microsoft.VisualStudio.Sdk.TestFramework.Tests/Mocks/GreetMock.cs (hunk 7 above)
Comment (murphymcquet-msft) [reply test/Microsoft.VisualStudio.Sdk.TestFramework.Tests/Mocks/GreetMock.cs:18]: Great, thank you. I was struggling a bit with how to invoke the call.
